modules/ctl/src/main/scala/hi/deploy.scala

```diff
@@ -0,0 +1,267 @@
+    gosub("Verifying Gem deploy image.") {
+      requireImage(s"$GemOrg/$GemImage:${dc.imageVersion}")
+    }
+
+  def fileContentsAtDeployCommit(cDeploy: DeployCommit, path: String): CtlIO[List[String]] =
+    if (cDeploy.uncommitted) fileContents(path)
+    else fileContentsAtCommit(cDeploy.commit, path)
+
+  def requireImage(nameAndVersion: String): CtlIO[Image] = {
+    info(s"Looking for $nameAndVersion") *>
+    findImage(nameAndVersion).flatMap {
+      case None    =>
+        warn(s"Cannot find image locally. Pulling (could take a few minutes).") *> pullImage(nameAndVersion).flatMap {
+          case None    => error(s"Image was not found.") *> exit(-1)
+          case Some(i) => info(s"Image is ${i.hash}") as i
+        }
+      case Some(i) => info(s"Image is ${i.hash}") as i
+    }
+  }
+
+  def getPostgresImage(cDeploy: DeployCommit): CtlIO[Image] =
+    gosub("Verifying Postgres deploy image.") {
+      fileContentsAtDeployCommit(cDeploy, "pg-image.txt").flatMap { ss =>
+        ss.map(_.trim).filterNot(s => s.startsWith("#") || s.trim.isEmpty) match {
+          case List(name) => requireImage(name)
+          case _          =>
+            error(s"Cannot determine Postgres image for deploy commit $cDeploy") *>
+            exit(-1)
+        }
+      }
+    }
+
+  def verifyLineage(base: DeployCommit, dc: DeployCommit, force: Boolean): CtlIO[Unit] =
+    gosub("Verifying lineage.") {
+      (base, dc) match {
+        case (DeployCommit(b, true),  DeployCommit(d, true))  if b === d => info("Base and deploy commits are matching and UNCOMMITTED. All is well.")
+        case (DeployCommit(b, false), DeployCommit(d, false)) if b === d =>
+          info("Base and deploy commits are identical.") *> {
+            if (force) info("--force was specified, so upgrading anyway.")
+            else       info("Nothing to do. Use --force to do upgrade anyway.") *> exit(0)
+          }
+        case (DeployCommit(b, _),     DeployCommit(d, _)) =>
+          isAncestor(b, d).flatMap {
+            case true  => info( s"Base commit is an ancestor of deploy commit (as it should be).")
+            case false => error(s"Base commit is not an ancestor of deploy commit.") *> exit[Unit](-1)
+          }
+      }
+    }
+
+  def ensureNoRunningDeployments: CtlIO[Unit] =
+    gosub("Ensuring that there is no running deployment.") {
+      findRunningContainersWithLabel("edu.gemini.commit").flatMap {
+        case Nil => info("There are no running deployments.")
+        case cs  =>
+          for {
+            c <- config
+            _ <- error(s"There is already a running deployment on ${c.server}")
+            _ <- exit[Unit](-1)
+          } yield ()
+      }
+    }
+
+  def createDatabaseContainer(nDeploy: Int, cDeploy: DeployCommit, iPg: Image, kPrev: Option[Container]): CtlIO[Container] =
+    gosub(s"Creating Postgres container from image ${iPg.hash}") {
+      for {
+        r <- isRemote // irritating … see below
```

Comment:
It's used below to decide whether or not to include quote marks around an argument, because in the remote cases the command is added to an `ssh` command and it needs different quoting for args that contain spaces. It's kind of hacky and can probably be fixed at a lower level. Originally I was treating local deploys uniformly and just `ssh`ing back to localhost but that seemed needlessly complicated so I got rid of that but now I have to deal with differences in quoting.